pkg/provision/cleanup.go

```diff
@@ -0,0 +1,250 @@
+	pools, err := ioutil.ReadDir("/mnt")
+	if err != nil {
+		return errors.Wrap(err, "failed to list available pools")
+	}
+
+	for _, pool := range pools {
+		if !pool.IsDir() {
+			continue
+		}
+		path := filepath.Join("/mnt", pool.Name())
+		subvols, err := utils.SubvolumeList(ctx, path)
+		if err != nil {
+			// A Pool might not be mounted here, skip it
+			log.Error().Err(err).Msgf("failed to list available subvolumes in: %s", path)
+			continue
+		}
+
+		qgroups, err := utils.QGroupList(ctx, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to list available cgroups in: %s", path)
+		}
+
+		for _, subvol := range subvols {
+
+			qgroup, ok := qgroups[fmt.Sprintf("0/%d", subvol.ID)]
+			if !ok {
+				log.Info().Msgf("skipping volume '%s' has no assigned quota", subvol.Path)
+				continue
+			}
+
+			// now, is this subvol in one of the toDelete ?
+			if _, ok := toDelete[filepath.Base(subvol.Path)]; ok {
+				// delete the subvolume
+				log.Info().Msgf("deleting subvolume %s", subvol.Path)
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+
+			// if the subvolume is a zdb and has 0 maxRfer, don't skip here. It might need to be deleted
+			zdbVolume := strings.HasPrefix(subvol.Path, "zdb") && qgroup.MaxRfer != 0
+
+			// we only handle volumes that are 256MiB or 10MiB
+			if qgroup.MaxRfer != 268435456 && qgroup.MaxRfer != 10485760 && !zdbVolume {
+				log.Info().Msgf("skipping volume '%s' is of size: %d", subvol.Path, qgroup.MaxRfer)
+				continue
+			}
+
+			// now, is this subvol in one of the toSave ?
+			if _, ok := toSave[filepath.Base(subvol.Path)]; ok {
+				log.Info().Msgf("skipping volume '%s' is used", subvol.Path)
+				continue
+			}
+
+			// delete the subvolume
+			log.Info().Msgf("deleting subvolume %s", subvol.Path)
+			if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+				log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+			}
+			if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+				log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+			}
+		}
+	}
+
+	return nil
+}
+
+// checks running containers for subvolumes that might need to be saved because they are used
+// and subvolumes that might need to be deleted because they have no attached container anymore
+func checkContainers(msgBrokerCon string) (map[string]struct{}, map[string]struct{}, error) {
+	toSave := make(map[string]struct{})
+	toDelete := make(map[string]struct{})
+
+	for _, ns := range lines(must(execute("ctr", "namespace", "ls", "-q"))) {
+		log.Info().Msgf("Checking namespace %s", ns)
+		for _, cnt := range lines(must(execute("ctr", "-n", ns, "container", "ls", "-q"))) {
```

Comment:
Is there any reason you use ctr CLI instead of talking directly to containerd?

Reply:
No not really, this is how azmy did it. Should I refactor this?

Reply:
I did it in a one time cleaner script that we need to run to clean up some nodes. I think now we can indeed use containerd directly.

Reply:
refactored :)